[JSC] Implement RegExp Duplicate Named Capture Groups

msaboff opened the pull request:
#### 9ed7d9c4e36677ce84c2bd2b2b225d240365fa19
<pre>
[JSC] Implement RegExp Duplicate Named Capture Groups
<a href="https://bugs.webkit.org/show_bug.cgi?id=252553">https://bugs.webkit.org/show_bug.cgi?id=252553</a>
rdar://100335581

Reviewed by Yusuke Suzuki.

This change implements RegExp Duplicate Named Capture Groups (<a href="https://github.com/tc39/ecma262/tree/duplicate-named-capture-groups).">https://github.com/tc39/ecma262/tree/duplicate-named-capture-groups).</a>
Duplicate named captures have a unique positive ID like subpatternIds.  When matching, either in the interpreter or JIT,
space for an unsgined is added in the &quot;output&quot; vector for each duplicate named capture.  When a subpattern that is one of the
duplicates for a named capture matches, that subpattern&apos;s ID is saved in its duplicate named capture location of the output
vector.  When such a subpattern doesn&apos;t match or backtrackes, a zero is written in the named capture&apos;s output location.
When either matching a backreference for a named capture or creating the match result, the participating subpattern for a
duplicate named capture is indirectly accessed via the named capture&apos;s output location.

YarrPatterns now contains a map from a duplicate named capture to a vector of unsigneds.  The first value in that vector is
that duplicate named group&apos;s unique ID.  The remaining values are the subpatternIds that are part of the named group.
There is also a vector were the value at subpatternId index contains its duplicate named group&apos;s ID.  These allow for
finding the subpatternId&apos;s for a given duplicate named capture group as well as finding the duplicated named capture ID
for a given subpatternId.

Updated existing LayoutTests/js/regexp-named-capture-groups.html to no longer fail with duplcxate capture names.
Enabled duplicate named capture tests in test262.

* JSTests/test262/config.yaml: Enabled regexp-duplicate-named-groups
* JSTests/stress/regexp-duplicate-named-captures.js: Added.
(arrayToString):
(objectToString):
(dumpValue):
(compareArray):
(compareGroups):
(testRegExp):
(testRegExpSyntaxError):
(testRegExp.x.a):
(testRegExp.a.x):
* LayoutTests/js/regexp-named-capture-groups-expected.txt: Removed test for syntax error with a duplicate group name.
* LayoutTests/js/script-tests/regexp-named-capture-groups.js: Removed test for syntax error with a duplicate group name.
* Source/JavaScriptCore/runtime/RegExp.cpp:
(JSC::RegExp::finishCreation):
* Source/JavaScriptCore/runtime/RegExp.h:
* Source/JavaScriptCore/runtime/RegExpInlines.h:
(JSC::RegExp::matchInline):
* Source/JavaScriptCore/runtime/RegExpMatchesArray.h:
(JSC::createRegExpMatchesArray):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::substituteBackreferencesSlow):
(JSC::replaceUsingRegExpSearch):
* Source/JavaScriptCore/yarr/Yarr.h:
* Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::ParenthesesDisjunctionContext):
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::restoreOutput):
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::getDisjunctionContext):
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::backupOffsetForDuplicateNamedGroup):
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::allocationSize):
(JSC::Yarr::Interpreter::allocParenthesesDisjunctionContext):
(JSC::Yarr::Interpreter::matchBackReference):
(JSC::Yarr::Interpreter::backtrackBackReference):
(JSC::Yarr::Interpreter::recordParenthesesMatch):
(JSC::Yarr::Interpreter::resetMatches):
(JSC::Yarr::Interpreter::parenthesesDoBacktrack):
(JSC::Yarr::Interpreter::matchParenthesesOnceBegin):
(JSC::Yarr::Interpreter::matchParenthesesOnceEnd):
(JSC::Yarr::Interpreter::backtrackParenthesesOnceBegin):
(JSC::Yarr::Interpreter::backtrackParenthesesOnceEnd):
(JSC::Yarr::Interpreter::matchParentheses):
(JSC::Yarr::Interpreter::backtrackParentheses):
(JSC::Yarr::ByteCompiler::compile):
(JSC::Yarr::ByteCompiler::atomBackReference):
(JSC::Yarr::ByteCompiler::atomParentheticalAssertionBegin):
(JSC::Yarr::ByteCompiler::atomParentheticalAssertionEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesSubpatternEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesOnceEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesTerminalEnd):
(JSC::Yarr::ByteCompiler::emitDisjunction):
(JSC::Yarr::ByteTermDumper::dumpTerm):
* Source/JavaScriptCore/yarr/YarrInterpreter.h:
(JSC::Yarr::ByteTerm::ByteTerm):
(JSC::Yarr::ByteTerm::ParentheticalAssertionBegin):
(JSC::Yarr::ByteTerm::ParentheticalAssertionEnd):
(JSC::Yarr::ByteTerm::containsAnyCaptures):
(JSC::Yarr::ByteTerm::subpatternId):
(JSC::Yarr::ByteTerm::duplicateNamedGroupId):
(JSC::Yarr::ByteTerm::firstSubpatternId):
(JSC::Yarr::ByteTerm::lastSubpatternId):
(JSC::Yarr::BytecodePattern::BytecodePattern):
(JSC::Yarr::BytecodePattern::hasDuplicateNamedCaptureGroups const):
(JSC::Yarr::BytecodePattern::offsetForDuplicateNamedGroupId):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
(JSC::Yarr::loadSubPatternIdForDuplicateNamedGroup):
(JSC::Yarr::loadSubPattern):
(JSC::Yarr::loadSubPatternEnd):
(JSC::Yarr::YarrGenerator):
(JSC::Yarr::compile):
(JSC::Yarr::compileInline):
(JSC::Yarr::dumpCompileFailure): Deleted.
(JSC::Yarr::jitCompile): Deleted.
(JSC::Yarr::jitCompileInlinedTest): Deleted.
* Source/JavaScriptCore/yarr/YarrJITRegisters.h:
* Source/JavaScriptCore/yarr/YarrParser.h:
(JSC::Yarr::Parser::NamedCaptureGroups::NamedCaptureGroups):
(JSC::Yarr::Parser::NamedCaptureGroups::contains):
(JSC::Yarr::Parser::NamedCaptureGroups::isEmpty):
(JSC::Yarr::Parser::NamedCaptureGroups::reset):
(JSC::Yarr::Parser::NamedCaptureGroups::nextAlternative):
(JSC::Yarr::Parser::NamedCaptureGroups::pushParenthesis):
(JSC::Yarr::Parser::NamedCaptureGroups::popParenthesis):
(JSC::Yarr::Parser::NamedCaptureGroups::add):
(JSC::Yarr::Parser::parseEscape):
(JSC::Yarr::Parser::parseParenthesesBegin):
(JSC::Yarr::Parser::parseParenthesesEnd):
(JSC::Yarr::Parser::parseTokens):
(JSC::Yarr::Parser::parse):
(JSC::Yarr::Parser::handleIllegalReferences):
(JSC::Yarr::Parser::containsIllegalNamedForwardReference):
(JSC::Yarr::Parser::resetForReparsing):
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::UnresolvedForwardReference::UnresolvedForwardReference):
(JSC::Yarr::YarrPatternConstructor::UnresolvedForwardReference::hasNamedGroup):
(JSC::Yarr::YarrPatternConstructor::UnresolvedForwardReference::namedGroup):
(JSC::Yarr::YarrPatternConstructor::namedCaptureGroupIdForName):
(JSC::Yarr::YarrPatternConstructor::tryConvertingForwardReferencesToBackreferences):
(JSC::Yarr::YarrPatternConstructor::atomParenthesesSubpatternBegin):
(JSC::Yarr::YarrPatternConstructor::atomParenthesesEnd):
(JSC::Yarr::YarrPatternConstructor::atomNamedBackReference):
(JSC::Yarr::YarrPatternConstructor::atomNamedForwardReference):
(JSC::Yarr::YarrPatternConstructor::disjunction):
(JSC::Yarr::YarrPatternConstructor::setupDuplicateNamedCaptures):
(JSC::Yarr::YarrPattern::compile):
* Source/JavaScriptCore/yarr/YarrPattern.h:
(JSC::Yarr::PatternTerm::ForwardReference):
(JSC::Yarr::YarrPattern::resetForReparsing):
(JSC::Yarr::YarrPattern::offsetVectorBaseForNamedCaptures const):
(JSC::Yarr::YarrPattern::offsetsSize const):
(JSC::Yarr::YarrPattern::offsetForDuplicateNamedGroupId):
(JSC::Yarr::YarrPattern::hasDuplicateNamedCaptureGroups const):
(JSC::Yarr::BackTrackInfoBackReference::backReferenceSizeIndex):

Canonical link: <a href="https://commits.webkit.org/260692@main">https://commits.webkit.org/260692@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4baebc7198e97b0049a1a64056e661fe71449b5d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109070 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18152 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41892 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/605 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118337 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19611 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9454 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101311 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114830 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14702 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97938 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42868 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29591 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84598 "Passed tests") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/98099 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10934 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/30937 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/98891 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/9025 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11673 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/31024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17047 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50536 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106565 "Built successfully") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7378 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13279 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26412 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -132,54 +133,98 @@ class Interpreter {
         {
             static_assert(alignof(ParenthesesDisjunctionContext) <= sizeof(void*));
-            size_t rawSize = sizeof(ParenthesesDisjunctionContext) - sizeof(unsigned) + (Checked<size_t>(numberOfSubpatterns) * 2U) * sizeof(unsigned);
+            size_t rawSize = sizeof(ParenthesesDisjunctionContext) +                 Checked<size_t>(numBackupIds) * sizeof(unsigned);
```
Remove unnecessary spaces.

msaboff replied:
Done.

Constellation commented on Source/JavaScriptCore/yarr/YarrJITRegisters.h:
```diff
@@ -68,7 +68,7 @@ struct YarrJITDefaultRegisters {
     static constexpr GPRReg remainingMatchCount = ARM64Registers::x9;
     static constexpr GPRReg regUnicodeInputAndTrail = ARM64Registers::x10;
-    static constexpr GPRReg unicodeTemp = ARM64Registers::x5;
+    static constexpr GPRReg unicodeAndSubpatternIdTemp = ARM64Registers::x5;
```
Now these registers are used in different purpose too. And they can be InvalidGPRReg if it is inlined RegExp.
Can you ensure that inlining does not happen when `unicodeAndSubpatternIdTemp` is used? Previously, we disabled Inlining when it has `unicode()`, but since this is used in different purpose, we should check the condition for that too.

msaboff replied:
In YarrJIT.cpp::YarrGenerator::compile() ~lines 4616-4622 we set "canInline" based on multiple conditions including having no subpatterns or back references.  The new use of unicodeTemp, now called unicodeAndSubpatternIdTemp, is for back references so the existing check is sufficient.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -2386,6 +2459,15 @@ class ByteCompiler {
+            if (duplicateNamedGroupId) {
+                m_bodyDisjunction->terms[endTerm].atom.parenIds.duplicateNamedGroupId = duplicateNamedGroupId;
+                m_bodyDisjunction->terms[beginTerm].atom.parenIds.duplicateNamedGroupId = duplicateNamedGroupId;
+
```
Remove this empty line.

msaboff replied:
Done.

msaboff committed:
[JSC] Implement RegExp Duplicate Named Capture Groups
https://bugs.webkit.org/show_bug.cgi?id=252553
rdar://100335581

Reviewed by Yusuke Suzuki.

This change implements RegExp Duplicate Named Capture Groups (https://github.com/tc39/ecma262/tree/duplicate-named-capture-groups).
Duplicate named captures have a unique positive ID like subpatternIds.  When matching, either in the interpreter or JIT,
space for an unsgined is added in the "output" vector for each duplicate named capture.  When a subpattern that is one of the
duplicates for a named capture matches, that subpattern's ID is saved in its duplicate named capture location of the output
vector.  When such a subpattern doesn't match or backtrackes, a zero is written in the named capture's output location.
When either matching a backreference for a named capture or creating the match result, the participating subpattern for a
duplicate named capture is indirectly accessed via the named capture's output location.

YarrPatterns now contains a map from a duplicate named capture to a vector of unsigneds.  The first value in that vector is
that duplicate named group's unique ID.  The remaining values are the subpatternIds that are part of the named group.
There is also a vector were the value at subpatternId index contains its duplicate named group's ID.  These allow for
finding the subpatternId's for a given duplicate named capture group as well as finding the duplicated named capture ID
for a given subpatternId.

Updated existing LayoutTests/js/regexp-named-capture-groups.html to no longer fail with duplcxate capture names.
Enabled duplicate named capture tests in test262.

* JSTests/test262/config.yaml: Enabled regexp-duplicate-named-groups
* JSTests/stress/regexp-duplicate-named-captures.js: Added.
(arrayToString):
(objectToString):
(dumpValue):
(compareArray):
(compareGroups):
(testRegExp):
(testRegExpSyntaxError):
(testRegExp.x.a):
(testRegExp.a.x):
* LayoutTests/js/regexp-named-capture-groups-expected.txt: Removed test for syntax error with a duplicate group name.
* LayoutTests/js/script-tests/regexp-named-capture-groups.js: Removed test for syntax error with a duplicate group name.
* Source/JavaScriptCore/runtime/RegExp.cpp:
(JSC::RegExp::finishCreation):
* Source/JavaScriptCore/runtime/RegExp.h:
* Source/JavaScriptCore/runtime/RegExpInlines.h:
(JSC::RegExp::matchInline):
* Source/JavaScriptCore/runtime/RegExpMatchesArray.h:
(JSC::createRegExpMatchesArray):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::substituteBackreferencesSlow):
(JSC::replaceUsingRegExpSearch):
* Source/JavaScriptCore/yarr/Yarr.h:
* Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::ParenthesesDisjunctionContext):
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::restoreOutput):
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::getDisjunctionContext):
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::backupOffsetForDuplicateNamedGroup):
(JSC::Yarr::Interpreter::ParenthesesDisjunctionContext::allocationSize):
(JSC::Yarr::Interpreter::allocParenthesesDisjunctionContext):
(JSC::Yarr::Interpreter::matchBackReference):
(JSC::Yarr::Interpreter::backtrackBackReference):
(JSC::Yarr::Interpreter::recordParenthesesMatch):
(JSC::Yarr::Interpreter::resetMatches):
(JSC::Yarr::Interpreter::parenthesesDoBacktrack):
(JSC::Yarr::Interpreter::matchParenthesesOnceBegin):
(JSC::Yarr::Interpreter::matchParenthesesOnceEnd):
(JSC::Yarr::Interpreter::backtrackParenthesesOnceBegin):
(JSC::Yarr::Interpreter::backtrackParenthesesOnceEnd):
(JSC::Yarr::Interpreter::matchParentheses):
(JSC::Yarr::Interpreter::backtrackParentheses):
(JSC::Yarr::ByteCompiler::compile):
(JSC::Yarr::ByteCompiler::atomBackReference):
(JSC::Yarr::ByteCompiler::atomParentheticalAssertionBegin):
(JSC::Yarr::ByteCompiler::atomParentheticalAssertionEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesSubpatternEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesOnceEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesTerminalEnd):
(JSC::Yarr::ByteCompiler::emitDisjunction):
(JSC::Yarr::ByteTermDumper::dumpTerm):
* Source/JavaScriptCore/yarr/YarrInterpreter.h:
(JSC::Yarr::ByteTerm::ByteTerm):
(JSC::Yarr::ByteTerm::ParentheticalAssertionBegin):
(JSC::Yarr::ByteTerm::ParentheticalAssertionEnd):
(JSC::Yarr::ByteTerm::containsAnyCaptures):
(JSC::Yarr::ByteTerm::subpatternId):
(JSC::Yarr::ByteTerm::duplicateNamedGroupId):
(JSC::Yarr::ByteTerm::firstSubpatternId):
(JSC::Yarr::ByteTerm::lastSubpatternId):
(JSC::Yarr::BytecodePattern::BytecodePattern):
(JSC::Yarr::BytecodePattern::hasDuplicateNamedCaptureGroups const):
(JSC::Yarr::BytecodePattern::offsetForDuplicateNamedGroupId):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
(JSC::Yarr::loadSubPatternIdForDuplicateNamedGroup):
(JSC::Yarr::loadSubPattern):
(JSC::Yarr::loadSubPatternEnd):
(JSC::Yarr::YarrGenerator):
(JSC::Yarr::compile):
(JSC::Yarr::compileInline):
(JSC::Yarr::dumpCompileFailure): Deleted.
(JSC::Yarr::jitCompile): Deleted.
(JSC::Yarr::jitCompileInlinedTest): Deleted.
* Source/JavaScriptCore/yarr/YarrJITRegisters.h:
* Source/JavaScriptCore/yarr/YarrParser.h:
(JSC::Yarr::Parser::NamedCaptureGroups::NamedCaptureGroups):
(JSC::Yarr::Parser::NamedCaptureGroups::contains):
(JSC::Yarr::Parser::NamedCaptureGroups::isEmpty):
(JSC::Yarr::Parser::NamedCaptureGroups::reset):
(JSC::Yarr::Parser::NamedCaptureGroups::nextAlternative):
(JSC::Yarr::Parser::NamedCaptureGroups::pushParenthesis):
(JSC::Yarr::Parser::NamedCaptureGroups::popParenthesis):
(JSC::Yarr::Parser::NamedCaptureGroups::add):
(JSC::Yarr::Parser::parseEscape):
(JSC::Yarr::Parser::parseParenthesesBegin):
(JSC::Yarr::Parser::parseParenthesesEnd):
(JSC::Yarr::Parser::parseTokens):
(JSC::Yarr::Parser::parse):
(JSC::Yarr::Parser::handleIllegalReferences):
(JSC::Yarr::Parser::containsIllegalNamedForwardReference):
(JSC::Yarr::Parser::resetForReparsing):
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::UnresolvedForwardReference::UnresolvedForwardReference):
(JSC::Yarr::YarrPatternConstructor::UnresolvedForwardReference::hasNamedGroup):
(JSC::Yarr::YarrPatternConstructor::UnresolvedForwardReference::namedGroup):
(JSC::Yarr::YarrPatternConstructor::namedCaptureGroupIdForName):
(JSC::Yarr::YarrPatternConstructor::tryConvertingForwardReferencesToBackreferences):
(JSC::Yarr::YarrPatternConstructor::atomParenthesesSubpatternBegin):
(JSC::Yarr::YarrPatternConstructor::atomParenthesesEnd):
(JSC::Yarr::YarrPatternConstructor::atomNamedBackReference):
(JSC::Yarr::YarrPatternConstructor::atomNamedForwardReference):
(JSC::Yarr::YarrPatternConstructor::disjunction):
(JSC::Yarr::YarrPatternConstructor::setupDuplicateNamedCaptures):
(JSC::Yarr::YarrPattern::compile):
* Source/JavaScriptCore/yarr/YarrPattern.h:
(JSC::Yarr::PatternTerm::ForwardReference):
(JSC::Yarr::YarrPattern::resetForReparsing):
(JSC::Yarr::YarrPattern::offsetVectorBaseForNamedCaptures const):
(JSC::Yarr::YarrPattern::offsetsSize const):
(JSC::Yarr::YarrPattern::offsetForDuplicateNamedGroupId):
(JSC::Yarr::YarrPattern::hasDuplicateNamedCaptureGroups const):
(JSC::Yarr::BackTrackInfoBackReference::backReferenceSizeIndex):

Canonical link: https://commits.webkit.org/260692@main
